src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/ResultExtensions.cs

```diff
@@ -0,0 +1,432 @@
+            this Result<TLocalOk, TError> result,
+            Result<TOtherOk, TError> otherResult,
+            Func<TLocalOk, TOtherOk, TResultOk> okCombiner)
+            where TError : ICombinable<TError>
+        {
+            if (result == null) throw new ArgumentNullException(nameof(result));
+            if (otherResult == null) throw new ArgumentNullException(nameof(otherResult));
+            if (okCombiner == null) throw new ArgumentNullException(nameof(okCombiner));
+
+            return result.Match(
+                fromOk: localOk => otherResult.Match(
+                    fromOk: otherOk => new Result<TResultOk, TError>(okCombiner(localOk, otherOk)),
+                    fromError: otherError => new Result<TResultOk, TError>(otherError)
+                    ),
+                fromError: localError => otherResult.Match(
+                    fromOk: otherOk => new Result<TResultOk, TError>(localError),
+                    fromError: otherError => new Result<TResultOk, TError>(localError.Combine(otherError))
+                    )
+                );
+        }
+
+        /// <summary>
+        /// Combine two <see cref="Result{TOk,TError}"/> values into a single value containing a tuple
+        /// </summary>
+        /// <remarks>
+        /// Works as a logical <c>AND</c> - if both inputs are OK, the output is OK;
+        /// if either input is an error, the output is an error.
+        /// </remarks>
+        /// <returns></returns>
+        public static Result<(TFirstOk First, TSecondOk Second), TError> With<TFirstOk, TError, TSecondOk>(
+            this Result<TFirstOk, TError> first,
+            Result<TSecondOk, TError> second)
+            where TError : ICombinable<TError>
+        {
+            if (first == null) throw new ArgumentNullException(nameof(first));
+            if (second == null) throw new ArgumentNullException(nameof(second));
+
+            return first.With(second, (v1, v2) => (v1, v2));
+        }
+
+        /// <summary>
+        /// Combine two <see cref="Result{TOk,TError}"/> values using the <see cref="With{TLocalOk,TError,TOtherOk,TResultOk}"/>
+        /// method, which combines their errors if either/both are in an error state.
+        /// </summary>
+        /// <remarks>
+        /// Intended to be used in LINQ query syntax, with a join clause equating to true. E.g.
+        /// <code>
+        ///     from v1 in result1
+        ///     join v2 in result2 on true equals true
+        ///     select (v1, v2)
+        /// </code>
+        /// </remarks>
+        public static Result<TResult, TError> Join<TOuter, TError, TInner, TKey, TResult>(
+            this Result<TOuter, TError> outer,
+            Result<TInner, TError> inner,
+            Func<TOuter, TKey> outerKeySelector,
+            Func<TInner, TKey> innerKeySelector,
+            Func<TOuter, TInner, TResult> resultSelector)
+            where TError : ICombinable<TError>
```

Comment:
I don't think we should be supporting this (mis)use of the LINQ syntax. Good for a thought experiment, but not for actual use.